File: python/19_Remove_Nth_Node_From_End_of_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} head
    # @param {integer} n
    # @return {ListNode}
    def removeNthFromEnd(self, head, n):
        post=head
        cur=head
        while n>0:
            post=post.__next__
            n-=1
        if not post:
            return cur.__next__
        while post.__next__:
            post=post.__next__
            cur=cur.__next__
        cur.next=cur.next.__next__
        return head

File: Java/19_Remove_Nth_Node_From_End_of_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode keep = head;
        ListNode runner = new ListNode(0);
        runner.next = keep;
        if(keep==null||n<=0)return keep;
        int slow=0;
        while(head!=null&&slow<n){
            head=head.next;
            slow++;
        }
        while(head!=null){
            runner = runner.next;
            head = head.next;
        }
        if(runner.next==keep){
            return keep.next;
        }else if(runner.next.next==null){
            runner.next=null;
        }else{
            runner.next = runner.next.next;
        }
        return keep;
    }
}